File: helm/relayforge/src/relayforge/api/metrics.py
from fastapi import APIRouter
from fastapi.responses import Response
from prometheus_client import Counter, Gauge, Histogram, generate_latest

router = APIRouter()

requests_total = Counter(
    "relayforge_requests_total", "HTTP requests processed by result", ["result"]
)
jobs_created_total = Counter("relayforge_jobs_created_total", "Delivery jobs created")
idempotency_conflicts_total = Counter(
    "relayforge_idempotency_conflicts_total", "Idempotency key conflicts"
)
kubernetes_errors_total = Counter(
    "relayforge_kubernetes_errors_total", "Kubernetes API errors by operation", ["operation"]
)
kubernetes_retries_total = Counter(
    "relayforge_kubernetes_retries_total", "Kubernetes client retries by reason", ["reason"]
)
request_duration_seconds = Histogram(
    "relayforge_request_duration_seconds", "HTTP request duration seconds", ["result"]
)
active_jobs = Gauge("relayforge_active_jobs", "Active delivery jobs observed by this pod")
oldest_active_job_seconds = Gauge(
    "relayforge_oldest_active_job_seconds", "Age in seconds of the oldest active delivery job"
)


def result_label(status_code: int) -> str:
    """Категория результата для metrics — ограниченный набор значений."""
    if status_code == 401:
        return "unauthorized"
    if status_code == 409:
        return "conflict"
    if status_code in (429, 503):
        return "backpressure"
    if 200 <= status_code < 300:
        return "success"
    if 400 <= status_code < 500:
        return "invalid"
    return "error"


def observe_request(status_code: int, duration: float) -> None:
    result = result_label(status_code)
    requests_total.labels(result=result).inc()
    request_duration_seconds.labels(result=result).observe(duration)


def inc_jobs_created() -> None:
    jobs_created_total.inc()


def inc_idempotency_conflict() -> None:
    idempotency_conflicts_total.inc()


def inc_k8s_error(operation: str) -> None:
    kubernetes_errors_total.labels(operation=operation).inc()


def inc_k8s_retry(reason: str) -> None:
    kubernetes_retries_total.labels(reason=reason).inc()


def set_active_jobs(count: int, oldest_seconds: float) -> None:
    active_jobs.set(count)
    oldest_active_job_seconds.set(oldest_seconds)


@router.get("/metrics")
async def metrics():
    return Response(generate_latest(), media_type="text/plain; version=0.0.4")

File: helm_golang/relayforge/internal/api/metrics.go
package api

import (
	"net/http"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

// Metrics registry with the SAME names as relayforge/api/metrics.py.
var (
	requestsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{Name: "relayforge_requests_total", Help: "HTTP requests processed by result"},
		[]string{"result"},
	)
	jobsCreatedTotal = prometheus.NewCounter(
		prometheus.CounterOpts{Name: "relayforge_jobs_created_total", Help: "Delivery jobs created"},
	)
	idempotencyConflictsTotal = prometheus.NewCounter(
		prometheus.CounterOpts{Name: "relayforge_idempotency_conflicts_total", Help: "Idempotency key conflicts"},
	)
	kubernetesErrorsTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{Name: "relayforge_kubernetes_errors_total", Help: "Kubernetes API errors by operation"},
		[]string{"operation"},
	)
	kubernetesRetriesTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{Name: "relayforge_kubernetes_retries_total", Help: "Kubernetes client retries by reason"},
		[]string{"reason"},
	)
	requestDurationSeconds = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{Name: "relayforge_request_duration_seconds", Help: "HTTP request duration seconds"},
		[]string{"result"},
	)
	activeJobs = prometheus.NewGauge(
		prometheus.GaugeOpts{Name: "relayforge_active_jobs", Help: "Active delivery jobs observed by this pod"},
	)
	oldestActiveJobSeconds = prometheus.NewGauge(
		prometheus.GaugeOpts{Name: "relayforge_oldest_active_job_seconds", Help: "Age in seconds of the oldest active delivery job"},
	)
)

func init() {
	prometheus.MustRegister(
		requestsTotal, jobsCreatedTotal, idempotencyConflictsTotal,
		kubernetesErrorsTotal, kubernetesRetriesTotal, requestDurationSeconds,
		activeJobs, oldestActiveJobSeconds,
	)
}

// resultLabel mirrors metrics.result_label.
func resultLabel(statusCode int) string {
	switch {
	case statusCode == 401:
		return "unauthorized"
	case statusCode == 409:
		return "conflict"
	case statusCode == 429 || statusCode == 503:
		return "backpressure"
	case statusCode >= 200 && statusCode < 300:
		return "success"
	case statusCode >= 400 && statusCode < 500:
		return "invalid"
	default:
		return "error"
	}
}

func observeRequest(statusCode int, durationSeconds float64) {
	result := resultLabel(statusCode)
	requestsTotal.WithLabelValues(result).Inc()
	requestDurationSeconds.WithLabelValues(result).Observe(durationSeconds)
}

func incJobsCreated()          { jobsCreatedTotal.Inc() }
func incIdempotencyConflict()  { idempotencyConflictsTotal.Inc() }
func incK8sError(operation string) {
	kubernetesErrorsTotal.WithLabelValues(operation).Inc()
}
func incK8sRetry(reason string) {
	kubernetesRetriesTotal.WithLabelValues(reason).Inc()
}

func setActiveJobs(count int, oldestSeconds float64) {
	activeJobs.Set(float64(count))
	oldestActiveJobSeconds.Set(oldestSeconds)
}

// metricsHandler serves /metrics in the prometheus text exposition format,
// mirroring metrics.router GET /metrics.
func metricsHandler() http.Handler {
	return promhttp.Handler()
}